DifferentiationInterface/ext/DifferentiationInterfaceReactantExt/onearg.jl

```diff
@@ -0,0 +1,36 @@
+struct ReactantGradientPrep{F,G} <: GradientPrep
+    compiled_function::F
+    compiled_gradient::G
+end
+
+function DI.prepare_gradient(f, rebackend::ReactantBackend, x)
+    xr = to_rarray(x)
+    gradient_closure(xr) = DI.gradient(f, rebackend.backend, xr)
+    compiled_function = @compile f(xr)
+    compiled_gradient = @compile gradient_closure(xr)
+    return ReactantGradientPrep(compiled_function, compiled_gradient)
+end
+
+function DI.gradient(f, prep::ReactantGradientPrep, ::ReactantBackend, x)
+    @compat (; compiled_gradient) = prep
+    xr = to_rarray(x)
+    return compiled_gradient(xr)
+end
+
+function DI.value_and_gradient(f, prep::ReactantGradientPrep, ::ReactantBackend, x)
+    @compat (; compiled_function, compiled_gradient) = prep
+    xr = to_rarray(x)
+    return compiled_function(xr), compiled_gradient(xr)
```

Comment:
for performance reasons these should not be separate compiled functions (e.g. you should precompile exactly what's going to be run)

Reply:
Ideally yes. In practice, DI has a single preparation operator `prepare_gradient` but 4 variants of the operator itself: `gradient`, `gradient!`, `value_and_gradient`, `value_and_gradient!`. This makes the user experience more pleasant, but it does incur a slight overhead because you might be overpreparing.

Reply:
Or do you mean I should compile a function that returns the tuple `value, gradient`?

Reply:
The latter (otherwise reactant will have a lot of ops computed unnecessarily )

Reply:
So I’m actually suggesting that we should over prepare even more!

Reply:
Let's do it!